tools/wptserve/wptserve/server.py

```diff
@@ -406,20 +424,118 @@ def handle_one_request(self):
         )
         t.start()
         return t
 
+    def _stream_ws_thread(self, stream_id, queue):
+        frame = queue.get(True, None)
+
+        rfile, wfile = os.pipe()
+        rfile, wfile = os.fdopen(rfile, 'rb'), os.fdopen(wfile, 'wb', 0)  # needs to be unbuffer for websockets
+        stream_handler = H2HandlerCopy(self, frame, rfile)
+
+        h2request = H2Request(stream_handler)
+        h2response = H2Response(stream_handler, h2request)
+
+        dispatcher = dispatch.Dispatcher(self.server.ws_doc_root, None, False)
+        if not dispatcher.get_handler_suite(stream_handler.path):
+            h2response.set_error(404)
+            h2response.write()
+            return
+
+        request_wrapper = _WebSocketRequest(stream_handler, h2response)
+
+        handshaker = WsH2Handshaker(request_wrapper, dispatcher)
+        try:
+            handshaker.do_handshake()
+        except HandshakeException as e:
+            self.logger.info('Handshake failed for error: %s', e)
+            h2response.set_error(e.status)
+            h2response.write()
+            return
+
+        # h2 Handshaker prepares the headers but does not send them down the
+        # wire. Flush the headers here.
+        h2response.write_status_headers()
+
+        request_wrapper._dispatcher = dispatcher
+
+        # TODO(arenevier): we need two threads:
+        # - one to handle the frame queue
+        # - one to handle the request (dispatcher.transfer_data is blocking)
+        # the alternative is to have only one (blocking) thread. That thread
+        # will call transfer_data. That would require a special case in
+        # handle_one_request, to bypass the queue and write data to wfile
+        # directly.
+        t = threading.Thread(
+            target=Http2WebTestRequestHandler._stream_ws_sub_thread,
+            args=(self, request_wrapper, stream_handler, queue)
+        )
+        t.start()
+
+        while not self.close_connection:
+            frame = queue.get(True, None)
+
+            if isinstance(frame, DataReceived):
+                wfile.write(frame.data)
+                if frame.stream_ended:
+                    raise NotImplementedError("frame.stream_ended")
+                    wfile.close()
+            elif frame is None or isinstance(frame, (StreamReset, StreamEnded, ConnectionTerminated)):
+                self.logger.debug('(%s - %s) Stream Reset, Thread Closing' % (self.uid, stream_id))
+                break
+
+        t.join()
+
+    def _stream_ws_sub_thread(self, request, stream_handler, queue):
+        dispatcher = request._dispatcher
+        dispatcher.transfer_data(request)
+
+        stream_id = stream_handler.h2_stream_id
+        with stream_handler.conn as connection:
+            try:
+                # TODO(arenevier): reset_stream is needed. If we use end_stream,
+                # the websocket will not close cleanly. Is it according to spec?
+                connection.reset_stream(stream_id)
```

Comment:
@bencebeky Do you know the answer to this question?

Reply:
This is not according to the spec.  According to https://tools.ietf.org/html/rfc8441#section-5, closing the HTTP/2 stream is like closing the TCP connection for regular WebSockets.

I was not aware of this bug, and I did not find any relevant issue on the Chromium bugtracker or any TODO in the source code.  Alas I did not find a test for this behavior either.  I suggest you file a bug at https://crbug.com/new and add a link to that in this comment.

Reply:
I have opened https://bugs.chromium.org/p/chromium/issues/detail?id=1151393

@bencebeky: since this is not according to the spec, and if chromium has a bug, maybe I should remove the TODO and use end_stream here?

Reply:
Hm actually that makes a lot of sense to me.  Also, thank you for filing the bug.